src/Scripts/Python/RekoProgramAPI.cs

```diff
@@ -160,71 +169,78 @@ public void SetUserProcedure(ulong linearAddress, string decl)
-            return UserProcedure(linearAddress).OutputFile;
+            return UserProcedure(addr).OutputFile;
         }
 
         public void SetUserProcedureOutputFile(
-            ulong linearAddress, string outputFile)
+            Address addr, string outputFile)
         {
-            UserProcedure(linearAddress).OutputFile = outputFile;
+            EnsureUserProcedure(addr).OutputFile = outputFile;
         }
 
         private bool TryConvertAddress(
-            ulong linearAddress, [NotNullWhen(true)] out Address? addr)
+            string sAddress, [NotNullWhen(true)] out Address? addr)
         {
-            try
-            {
-                addr = Addr(linearAddress);
-                return true;
-            }
-            catch
-            {
-                addr = null;
-                return false;
-            }
-        }
-
-        private Address Addr(ulong linearAddress)
-        {
-            return program.SegmentMap.MapLinearAddressToAddress(linearAddress);
+            return program.Architecture.TryParseAddress(sAddress, out addr);
         }
 
-        private EndianImageReader CreateImageReader(ulong linearAddress)
+        private EndianImageReader CreateImageReader(Address addr)
         {
-            var addr = Addr(linearAddress);
             return program.CreateImageReader(program.Architecture, addr);
         }
 
         private IEnumerable<T> ReadData<T>(
             Func<EndianImageReader, T> reader,
-            ulong iStartAddr,
+            Address startAddr,
             long length)
         {
-            var rdr = CreateImageReader(iStartAddr);
+            var rdr = CreateImageReader(startAddr);
             var offStart = rdr.Offset;
             while (rdr.Offset - offStart < length)
             {
                 yield return reader(rdr);
             }
         }
 
-        private UserProcedure UserProcedure(ulong linearAddress)
+        private UserProcedure UserProcedure(Address addr)
+        {
+            TryGetUserProcedure(addr, out var userProc);
+            return userProc;
+        }
+
+        private UserProcedure EnsureUserProcedure(Address addr)
+        {
+            if (!TryGetUserProcedure(addr, out var userProc))
+                program.User.Procedures[addr] = userProc;
+            return userProc;
+        }
+
+        /// <summary>
+        /// Gets the user procedure definition at specified address or default
+        /// user procedure based on data discovered by scanner.
+        /// Returns true if there is user procedure definition at specified
+        /// address; otherwise, false
+        /// </summary>
+        private bool TryGetUserProcedure(Address addr, out UserProcedure userProc)
         {
-            var addr = Addr(linearAddress);
-            return program.User.Procedures[addr];
+            if (program.User.Procedures.TryGetValue(addr, out userProc))
+                return true;
+            var proc = program.Procedures[addr];
```

Comment:
I would have expected the out parameter `userProc` to be null here, and have the caller make the `UserProcedure`.  Also, calling `program.Procedures[addr]` will crash if no procedure exists at address `addr`; is this the intention?

Reply:
The out parameter userProc to be null after 2408f40.
> calling program.Procedures[addr] will crash if no procedure exists at address addr; is this the intention

Yes, this is the intention. I have done explicit throwing of exception at  2408f40 to make it clear.